src/Shared/CopyOnWriteDictionary.cs

```diff
@@ -80,11 +91,24 @@ protected CopyOnWriteDictionary(SerializationInfo info, StreamingContext context
 
             object comparer = info.GetValue(nameof(Comparer), typeof(IEqualityComparer<K>));
 
-            var b = ImmutableDictionary.Create<K, V>((IEqualityComparer<K>)comparer);
+            var b = GetInitialDictionary((IEqualityComparer<K>)comparer);
 
             _backing = b.AddRange((KeyValuePair<K, V>[])v);
         }
 
+        private static ImmutableDictionary<K, V> GetInitialDictionary(IEqualityComparer<K>? keyComparer)
+        {
+#if NET35
+            return ImmutableDictionary.Create<K, V>(keyComparer);
+#else
+            return typeof(K) == typeof(string) && keyComparer is MSBuildNameIgnoreCaseComparer
```

Comment:
nits:
- Make `MSBuildNameIgnoreCaseComparer` sealed to future-proof the is-a check?
- `typeof(K) == typeof(string)` looks redundant, there's no way `K` could be something else if `keyComparer` is `MSBuildNameIgnoreCaseComparer`

Reply:
I tried and failed to convince the type system that that `typeof` was unnecessary. Ideally we'd run this logic only when `K` is `string` but I'm not aware of a way to do that.

Reply:
I want to be able to say `string.Empty is K`, but that would match with object, IComparable, etc...

Reply:
Sealing `MSBuildNameIgnoreCaseComparer` breaks the build--I guess because the compiler currently doesn't know that passing a `MSBuildNameIgnoreCaseComparer` implies that `typeof(K) == typeof(string)`? When sealed, I get

> `error CS0030: Cannot convert type 'Microsoft.Build.Collections.MSBuildNameIgnoreCaseComparer' to 'System.Collections.Generic.IEqualityComparer<K>'`

What I'd really like is a specialized implementation of this for K==string, but there's no way to do so that I'm aware of.

Reply:
Hmm, though there may be another option. Looks like the only places we use CopyOnWriteDictionary with `K != string` are in tests . . . maybe just drop the generic?